fix: merge generator merges on absent mergekey  #12849

girstenbrei opened the pull request:
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

Closes #12849

crenshaw-dev commented:
I'm excited about the refactor, this code has needed streamlining. 

I'm a little worried about the bugfix. If I understand correctly, this change _could_ delete folks' Applications if they're relying on the erroneous behavior.

Do we need to put the behavior change behind a flag? Or is this a far-edge enough case that we can get away with just a warning in the release notes?

girstenbrei commented:
Good question: so if someone currently relies on this, this is what happens:

1. A non existing merge key (or a combination of such) leads to parameters getting indexed with empty values
2. The merge is performed between all parameters with empty values as the index

The result is that if the base generators parameters don't contain the merge keys, it will be combined with all other generated parameters that don't contain the merge keys. The result is the exact opposite of merging: all resulting parameter sets matching the merge keys are filtered from the result (their merge index is not empty).

But, if the base generators parameters now suddenly changes and starts to contain the merge keys, the behavior flips: Now the merge index is not empty, suddenly matching the correct additional parameter sets, "inverting" the generated set of values.

Someone could rely on this (as there is no 'filter' generator), but to do so (and avoid the accidental sudden match from above), they'd need to explicitly set a merge key which is never generated "by accident".
Someone also could rely on this by accident, having this in prod and just not knowing. But then what they have in prod is probably not what they intended to have in prod.

IMHO I don't really know. I'm just a single user, I don't know about the creativity of others and their willingness to go on an adventure :grin: . But as this is undocumented behavior (I'm aware of [Hyrums Law](https://www.hyrumslaw.com/) ), I think this mechanism is highly confusing and dangerous (what if someone didn't consider the sudden switch when the merge key starts to exist).
The average user (hopefully) experiences this behavior by accident (because it generates the opposite of their intention) and then change their ApplicationSet to produce the intended result. So I think of this as an edge case which would need deliberate ingenuity to explicitly don't do what is intended.

A warning definitely seems to be in order. If you think a feature is appropriate, I'm fine with that, too. Shouldn't be complicated now with the code cleaned up a little , I'm taking the win :grin:

girstenbrei commented:
Seconds later, another Idea

Not trying to make this complicated, but glancing at #12837 for a kind of compromise: We could also put this behind the merge strategy. Keep the old  behavior as `add_and_override_or_filter` and put the new one behind a new `add_and_override` or sth like that (the names are the best I can come up with right now, happily taking other suggestions).

On the other hand, this would 'commit' this confusing behavior to basically be official now (or can this somehow be marked deprecated?), so I don't know about that.

crenshaw-dev commented:
@girstenbrei I will come back to this, just have a lot going on atm :-)

crenshaw-dev commented:
> On the other hand, this would 'commit' this confusing behavior to basically be official now

Agreed. I'd rather just yeet the confusing behavior.

I'd suggest just adding a release note about the change. Would you mind adding that?

girstenbrei commented:
Hi there!
I noticed a wrongly named test and fixed that as well. Additionally I added one sentence into the docs explaining the merge generator behavior.

I wrote a note in the upgrade notes, I'm doing this for the first time so if you want it somewhere else or need a shorter explanation in other release notes, please let me know where, happy to add that.

Greets,
Chris
